Istio compat - use kub service ip for upstreams

mkazin opened the pull request:
# Description

This change includes support for setting endpoints based on service ip addresses rather than dns names, which is required for integration with Istio.

This required breaking up a large and complicated function which was making code changes difficult. 

# Context
This is phase two of the work required to prepare Edge for full Istio integration (see s-p 3461)
It follows the work done in #6196 , utilizing the helm value established there to enable this code change.

# Testing
This was tested manually by repeating the method described in the earlier work to enable injection ([see here](https://github.com/solo-io/gloo/pull/6312#issuecomment-1098467998)), along with steps [described here](https://github.com/murphye/demo-gloo-edge-istio) to generate an error from the petstore sevice when strict mTLS is enabled .

A config dump shows the desired endpoint configuration:
```

 "endpoint_config": {
  "@type": "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment",
  "cluster_name": "kube-svc:petstore-petstore-8080_petstore-16672974176135880674",
  "endpoints": [
   {
    "locality": {},
    "lb_endpoints": [
     {
      "endpoint": {
       "address": {
        "socket_address": {
         "address": "10.40.0.174",
         "port_value": 8080
        }
       },
       "health_check_config": {}
      },
      "health_status": "HEALTHY",
      "metadata": {
       "filter_metadata": {
        "envoy.lb": {
         "service": "petstore"
        }
       }
      },
      "load_balancing_weight": 1
     }
    ]
   }
```

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [X] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [X] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/6195

nfuden commented on projects/gloo/pkg/plugins/kubernetes/eds.go:
```diff
@@ -257,55 +265,87 @@ func filterEndpoints(
-					}
-					key := Epkey{addr.IP, port, podName, podNamespace, usRef}
+
+				if filterToIpAddresses {
```
Filtertoipaddresses doesnt mean much to me. Do you think we can either change the name to imply that its due to hitting services or just putting a comment here?

mkazin replied:
As I mentioned elsewhere, I'm switching this name to indicate the purpose of istio integration rather the implementation details, adding a comment to explain the required implementation.

nfuden commented on projects/gloo/pkg/plugins/kubernetes/eds.go:
```diff
@@ -257,55 +265,87 @@ func filterEndpoints(
+	return warnings
+}
+
+func findPortForEndpointSubset(subset kubev1.EndpointSubset, singlePortService bool, kubeServicePort *kubev1.ServicePort) uint32 {
```
findFirstPortInEndpointSubsets?

mkazin replied:
Much better!

nfuden reviewed (commented):
Do we think we can have a test with this os env or is that a large lift?

nfuden commented on projects/gloo/pkg/plugins/kubernetes/eds.go:
```diff
@@ -203,50 +207,54 @@ func (c *edsWatcher) watch(writeNamespace string, opts clients.WatchOpts) (<-cha
+		if svc.Namespace != spec.GetServiceNamespace() || svc.Name != spec.GetServiceName() {
+			continue
+		}
+		if len(svc.Spec.Ports) == 1 {
+			if spec.GetServicePort() == uint32(svc.Spec.Ports[0].Port) {
+				return &svc.Spec.Ports[0], true
+			}
+		}
+		for _, port := range svc.Spec.Ports {
+			if spec.GetServicePort() == uint32(port.Port) {
+				return &port, false
+			}
+		}
```
Since we are already reframing this code we could consider calling out the fact that we return whether this is a single port service.
Either as a method comment and/or doing something like the following

```suggestion
		for _, port := range svc.Spec.Ports {
			if spec.GetServicePort() == uint32(port.Port) {
				return &port,  (len(svc.Spec.Ports) == 1)
			}
		}
```

nfuden commented on install/helm/gloo/templates/1-gloo-deployment.yaml:
```diff
@@ -221,6 +221,10 @@ spec:
             value: {{ .Values.gloo.logLevel }}
         {{- end}}
+        {{- if .Values.global.istioIntegration.enableIstioSidecarOnGateway }}
+          - name: ENABLE_ISTIO_SIDECAR
```
nit: For future reusability of this feature (say if this is not only triggerable on istio) I would love a name for the env var to be descriptive of this change. Say KUBE_IP_FOR_SERVICES (preferably better than that but cant think of a good one off the top of my head)

mkazin replied:
I don't know if getting more specific about the details of implementation is the right direction. Should more code changes become necessary to support istio, that name would become obsolete.
For the same reason, I'm also removing the "sidecar" mention (which isn't part of the code implementation in eds).
Rather, I'm suggesting ENABLE_ISTIO_INTEGRATION.

mkazin commented:
> Do we think we can have a test with this os env or is that a large lift?

Not a comprehensive test to verify the ip/hostname translation. Given no real testing on this module that'd take more time than I want to spend now.

I've extracted  the environment variable logic into its own method and added the fairly trivial testing it could use.

murphye commented on changelog/v1.12.0-beta13/istio-ip-routing.yaml:
```diff
@@ -0,0 +1,5 @@
+- type: NEW_FEATURE
+  issueLink: https://github.com/solo-io/gloo/issues/6195
+  resolvesIssue: true
+  description: Add support for enabling ip routing when injecting istio sidecar.
```
I would be more descriptive here just for reference. Routing to service IP rather than pod IP

mkazin replied:
Thanks- how's this new one I changed it to?
> Add support for istio integration in endpoint discovery, routing to service IP rather than pod IP.

ben-taussig-solo reviewed (commented):
Looks good to me, waiting for all comments to be resolved and then I'm happy to sign off on this :)

mkazin commented:
/kick build-bot
different known flake- [Access Log in memory Access Logs Grpc](https://storage.googleapis.com/solo-public-build-logs/log-21f6849a-ce2d-423b-9d66-ee38890ada8b.txt)
